### DISCUSSION TITLE
Create custom frames

### DISCUSSION THREAD

**Discussion by fredvs** [mse-org/mseide-msegui#78] [General]
To create custom frames, you may use the property frame/frameimage and select a image-list of 8 bitmaps.

![msegui_roundcourner](https://github.com/mse-org/mseide-msegui/assets/3421249/84f749fd-30b6-4f65-bb3b-af847305607c)

The first is the top left corner, then the left side-line, then the bottom left corner, then bottom line, then right bottom corner, right side-line, right top corner and finally top side line.

Ok, it works but could it be possible to create each bitmap "live", with drawline-arc, etc and then assign it to each bitmap of the image list?
That would be great to not need to import images in the image list but create at runtime each part of the frame (corners and side-lines)

_Originally posted by @fredvs in https://github.com/mse-org/mseide-msegui/discussions/74#discussioncomment-8651715_

**Comment by fredvs** [comment]
Here demo how to assign a bgrabitmap to a maskedbitmap:
https://github.com/mse-org/mseide-msegui/discussions/68#discussioncomment-8659403

**Comment by fredvs** [comment]
Now only need to learn how to design the 8 bgrabitmaps of nice corners an sides with shadow and assign it to to each of the msegui images list.

**Comment by rchastain** [comment]
Thank you Fred.

> Ok, it works but could it be possible to create each bitmap "live", with drawline-ellipse, etc and then assign it to each bitmap of the image list?

As a challenge and as an exercise, why not? But (honestly) I think it has not much practical interest. An application shouldn't recreate its buttons style each time it is executed.

**Reply by fredvs** [reply]
>An application shouldn't recreate its buttons style each time it is executed.


Indeed if there is only one "skin of frame".
But if the app proposes many skins, with the "classical" way you need to add for each skin 8 bitmaps as resource.
And also, so the corners are "hard coded" and not customizable by the user.

But I agree it is more for challenge and to demonstrate the infinite possibilities of MSEgui.

;-)

**Comment by rchastain** [comment]
> 8 images that you have to create with Gimp or with LazPaint is a boring work (for me).

Same for me. I was rather thinking of something like this (very basic example to explain my idea).

[framecreator-01.zip](https://github.com/mse-org/mseide-msegui/files/14545689/framecreator-01.zip)

**Reply by fredvs** [reply]
I will test it tonight, write you later.

**Reply by fredvs** [reply]
Hello Roland.

OK, tested, nice project and nice idea.

**Comment by rchastain** [comment]
Not yet found how to use the generated images. Here is my attempt, if you want to take a look. The frame always stays red...

[demobutton-01.zip](https://github.com/mse-org/mseide-msegui/files/14545901/demobutton-01.zip)

**Reply by fredvs** [reply]
Hello Roland.

Here your demo updated.
It uses a facelist.
[demobutton_updated.zip](https://github.com/mse-org/mseide-msegui/files/14546137/demobutton_updated.zip)

You may take a look at tbutton1.frame.frameface.list.**offsetclick** and **offsetmouse**.

![facelist_param1](https://github.com/mse-org/mseide-msegui/assets/3421249/3d3d4af0-679b-446e-ac91-1068deb72b6b)

Also, for **tfacelist1**, look at list.**offset** (to determine where the first frame starts).

![facelist_param2](https://github.com/mse-org/mseide-msegui/assets/3421249/6e73e99b-2225-4a60-96cd-2dd650d9d409)


PS: I suppose that you are using a old version of MSEide to create the project because the color of warning is still red and error is yellow. If you are using last MSEide, this is not ok and must be fixed (but strange because it should be already fixed).

**Reply by rchastain** [reply]
Bravo Fred! It works indeed. Thanks for the hints and screenshots.

Yes, I have several versions of MSEide installed.

**Comment by fredvs** [comment]
There is a demo how to draw custom frame with BGRABitmap here https://github.com/mse-org/mseide-msegui/discussions/68#discussioncomment-8725987

**Reply by rchastain** [reply]
There are interesting things in that demo, but you will agree (I believe) that it is experimental code. If things were well done, we shouldn't have to use a timer.

**Reply by fredvs** [reply]
>If things were well done, we shouldn't have to use a timer.

Yes, of course it can be done better and without a timer, the timer is there to show a longer face-execute stroke.
Without a timer, this also works but the face appears for less time.

Note that these demos are experiments to give keys to how to manage BGRABitmap with MSEgui.

**Reply by fredvs** [reply]
Note also that this demo was done before to investigate deeper tfacelist.
With tfacelist you may also assign a face for each mouse-action (it is on my todo list).
On the demo, only the frame is assigned with tfacelist, the same should be done for the face.

**Reply by rchastain** [reply]
Not clear for me why tfacelist is needed here. But anyway thanks for the info.

**Reply by fredvs** [reply]
facelist is needed if you want to assign different frame-face when the button is not focused, if the button is hoovered and if the button is clicked.

tfacelist.list[0]  --> button is not focused,
tfacelist.list[1]  --> button is hoovered,
tfacelist.list[2]  --> button is clicked,

In the demo, facelist is used only to assign a frame when button is hoovered (green rounded-rectangle) and when button is clicked (red rounded-rectangle).

**Reply by fredvs** [reply]
Re-hello.

In the demo, the orange-face when clicking  is done on **tbutton.onexecute**.

Of course, this is not the best solution now that we have discovered **tfacelist**.
This should be done using **tfacelist.list[2] --> button is clicked** and assign the orange-face there.

Will be done asap.
[EDIT] Done: https://github.com/mse-org/mseide-msegui/discussions/78#discussioncomment-8730509

**Comment by rchastain** [comment]
By the way I found this hint by Martin. Maybe interesting to try?

```
Hint: Build MSEgui with -dmse_flushgdi to switch off double buffering to see
redrawing of screen. With addition of -dmse_slowdrawing the effect is even
more visible.
```

**Reply by fredvs** [reply]
So much to discover...

Interesting indeed.
Will try (asap).

**Reply by fredvs** [reply]
I did try.
With -**dmse_flushgdi** I dont feel difference.
With -**dmse_slowdrawing**  this is indeed much slower drawing, maybe only useful to debug.

**Comment by fredvs** [comment]
Hello.

Here demo BGRABitmap for face and frame using face/frame properties of  tfacelist.
This dont use any timer.
[EDIT] The face come back to not-clicked when the button is released (thanks Martin!)


[bgrabitmap_button_hoover_frame_face.zip](https://github.com/mse-org/mseide-msegui/files/14547116/bgrabitmap_button_hoover_frame_face.zip)

![bgra_button_frame2](https://github.com/mse-org/mseide-msegui/assets/3421249/2f1c75cd-52d6-4603-b1d3-3c2a4e0c3645)

**Reply by fredvs** [reply]
This **facelist** by action is very great.
Little bemol, it dont allow to change caption value of the button or caption position (or I did not find the way, even using templates).
Fortunately, this could be done by BGRABitmap and center the text where you want, of course without using Tbutton.caption.

**Reply by fredvs** [reply]
Allez, the last, with BGRABitmap caption and moving caption:
[EDITED] zip and gif updated.

[bgrabitmap_button_hoover_frame_face_text.zip](https://github.com/mse-org/mseide-msegui/files/14547988/bgrabitmap_button_hoover_frame_face_text.zip)

![bgra-button-frame-face-text](https://github.com/mse-org/mseide-msegui/assets/3421249/980583c6-43a4-4413-8135-250a47142494)

And without frame it is nice also (with much less code):

[bgrabitmap_button_hoover_face_text.zip](https://github.com/mse-org/mseide-msegui/files/14547963/bgrabitmap_button_hoover_face_text.zip)

![bgra-button-face-text](https://github.com/mse-org/mseide-msegui/assets/3421249/de5327b3-905d-4d78-8b70-73bbb2f02a76)

I think I've done the turn.
